File: backend/app/routes/learning.py
from fastapi import APIRouter, Depends, status, Query
from sqlalchemy.orm import Session
from app.database.session import get_db
from app.services.learning_service import LearningService

router = APIRouter()

@router.get("/questions/{id}/stage", status_code=status.HTTP_200_OK)
def get_hint_stage(id: int, student_id: int = Query(..., description="The ID of the student"), db: Session = Depends(get_db)):
    service = LearningService(db)
    return service.get_reveal_stage(user_id=student_id, question_id=id)

@router.post("/attempts/{attempt_id}/feedback", status_code=status.HTTP_200_OK)
def get_ai_feedback(attempt_id: int, student_id: int = Query(..., description="The ID of the student"), db: Session = Depends(get_db)):
    service = LearningService(db)
    feedback_text = service.generate_ai_feedback(user_id=student_id, attempt_id=attempt_id)
    return {"status": "success", "attempt_id": attempt_id, "feedback": feedback_text}

File: backend/app/routes/learning.js
const express = require('express');
const router = express.Router();
const LearningService = require('../services/learning_service');

router.get('/questions/:id/stage', async (req, res, next) => {
  try {
    const id = parseInt(req.params.id, 10);
    const studentId = parseInt(req.query.student_id, 10);

    if (isNaN(studentId)) {
      return res.status(400).json({
        error: {
          code: "VALIDATION_ERROR",
          message: "student_id is required as query parameter"
        }
      });
    }

    const stageDetails = await LearningService.getRevealStage(studentId, id);
    res.json(stageDetails);
  } catch (err) {
    next(err);
  }
});

router.post('/attempts/:attempt_id/feedback', async (req, res, next) => {
  try {
    const attemptId = parseInt(req.params.attempt_id, 10);
    const studentId = parseInt(req.query.student_id, 10);

    if (isNaN(studentId)) {
      return res.status(400).json({
        error: {
          code: "VALIDATION_ERROR",
          message: "student_id is required as query parameter"
        }
      });
    }

    const feedbackText = await LearningService.generateAiFeedback(studentId, attemptId);
    res.json({
      status: "success",
      attempt_id: attemptId,
      feedback: feedbackText
    });
  } catch (err) {
    next(err);
  }
});

module.exports = router;
